### PR TITLE
Scatter estimator in ParallelPostFit

### LINKED ISSUE #842: ParallelPostFit excessive scheduler memory use and CancelledError
Notebook with MCVE and all notes: https://nbviewer.jupyter.org/gist/rikturr/43336377678018d01d4f21f19dd7ef11

When using `ParallelPostFit` to train with pandas/numpy objects then predict on dask objects, I noticed that the scheduler memory use runs extremely high. Many times I would get a `CancelledError` and scheduler dying when calling `.predict()` with pretty small data sizes (refer to notebook for full code with outputs):

```python
X_train, X_test, y_train, y_test = ...
rf = ParallelPostFit(
    RandomForestClassifier(n_estimators=100, random_state=seed, n_jobs=-1)
)
_ = rf.fit(X_train, y_train)

preds = rf.predict(X_test)
_ = preds.compute()  # failure happens on this line after ~40 minutes
```

The scheduler memory balloons to npartitions of `X_test` * size of `rf`, which can get into the multiple GBs very fast. I noticed that each time an operation would get called on `preds`, this memory exchange would happen again. I realize that this is because `ParallePostFit` uses `map_partitions` behind the scenes, but does not broadcast the model objects. This causes Dask to send the object through the scheduler to each worker _each time you do something with `preds`_ (unless of course you persisted it).

Workaround is to broadcast the model object then use `map_partitions` yourself instead of the `ParallelPostFit` wrapper:

```python
rf = RandomForestClassifier(n_estimators=100, random_state=seed, n_jobs=-1)
_ = rf.fit(X_train, y_train)

rf_fut = client.scatter(rf, broadcast=True)

def dask_predict(df, model):
    return model.predict(df)

preds = X_test.map_partitions(
    dask_predict,
    model=rf_fut,
    meta=np.array([1])
)
```

I plan to follow up with a PR to fix this in the `ParallelPostFit` class

### PR DESCRIPTION
**Pull request by rikturr**
closes #842 

Functionality of ParallelPostFit is the same, except behind the scenes it now scatters the scikit-learn estimator if running with a distributed cluster. Compared to existing functionality showcased in #842 where scheduler memory increases by npartitions * estimator size, the memory increase is now just estimator size for a quick blip while it gets shipped out to the workers

Screen Shot 2021-07-02 at 1 21 26 PM

### CODE REVIEW

**Comment by mrocklin** [comment]
Thank you for tracking this down @rikturr .  A couple of questions/comments

1.  Does this still work if things are run in async mode?  (if they worked before, that is)
2. Do you have thoughts on how to test this?
3. You might want to try wrapping the estimator in dask.delayed instead.  That should handle the deduplication.  Actually, this is already an optimization in map_partitions and map_blocks.  We wrap keywords with large sizeof values in dask.delayed in order to avoid duplication.  It might be wise to track down why that didn't occur.  That might lead you to a more general fix.

**Comment by rikturr** [comment]
@mrocklin 
1. Looking into the docs for async mode and will report back
2. Broadly, I was thinking that a good way to test this and potentially other issues is to have some integration tests set up that would allow for more/larger tests with distributed clusters. Is something like that available or in the works? For this specific situation, it can be reproduced with a `LocalCluster` so should be able to cover in unit tests. I'm not sure if there's a way to get scheduler memory use programatically from dask? I think that could be a good way to verify that dups aren't being sent
3. Ah, I didn't realize that map_partitions and map_blocks already check object sizes but that makes a lot of sense. So the better solution would be to make sure that the estimator does get delayed so we don't have to change anything in dask-ml (or xgboost, or lightgbm, etc). Looks like sys.getsizeof() on scikit-learn estimators doesn't return the accurate size. Checking the size with pickle does though:
image
Do you think its reasonable to have map_partitions and map_blocks use pickle+getsizeof to get the proper size? Might introduce some latency. Perhaps there's another way to get a more accurate size

**Comment by mrocklin** [comment]
I recommend that you identify where this should have triggered within
map_partitions/map_blocks, construct a simple case showing that it doesn't,
and then raise an issue in dask/dask about the problem.  That tends to get
more eyeballs on it than dask/dask-ml

**Comment by rikturr** [comment]
closing in favor of fix for root issue: https://github.com/dask/dask/issues/7860
